source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -140,20 +235,15 @@ Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMa
 }
 
 Http::FilterDataStatus CompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
-  if (!skip_compression_) {
-    config_->stats().total_uncompressed_bytes_.add(data.length());
-    compressor_->compress(data, end_stream ? Envoy::Compression::Compressor::State::Finish
-                                           : Envoy::Compression::Compressor::State::Flush);
-    config_->stats().total_compressed_bytes_.add(data.length());
-  }
-  return Http::FilterDataStatus::Continue;
+  return maybeCompress(response_compressor_, config_->responseDirectionConfig().stats(), data,
+                       end_stream);
 }
 
 Http::FilterTrailersStatus CompressorFilter::encodeTrailers(Http::ResponseTrailerMap&) {
-  if (!skip_compression_) {
+  if (response_compressor_ != nullptr) {
```

Comment:
I understand now why this is needed here vs not in decompression. In decompression information about the end of the compressed stream is in the stream itself; vs. here the information needs to be external. This function could also use maybeCompress passing `true` as the last param. Also https://github.com/envoyproxy/envoy/issues/6894 will affect the request side as is, right? i.e we need to add the same logic to decodeTrailers.

Reply:
> Also #6894 will affect the request side as is, right? i.e we need to add the same logic to decodeTrailers.

That's a very good catch! Thanks. Added `decodeTrailers()`.